k8s/pkg/proxyregistryserver/registry.go

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2019 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package proxyregistryserver - forwards k8s commands from local registry to remote one
+package proxyregistryserver
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/networkservicemesh/networkservicemesh/pkg/tools/spanhelper"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/serviceregistry"
+
+	"github.com/pkg/errors"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"golang.org/x/net/context"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/clusterinfo"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/nsmd"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/registry"
+)
+
+const (
+	// NSRegistryForwarderLogPrefix - log prefix
+	NSRegistryForwarderLogPrefix = "Network Service Registry Forwarder"
+	// NSMRSAddressEnv - environment variable - address of Network Service Registry Server to forward NSE registry requests
+	NSMRSAddressEnv = "NSMRS_ADDRESS"
+	// NSMRSReconnectInterval - reconnect interval to NSMRS if connection refused
+	NSMRSReconnectInterval = 15 * time.Second
+)
+
+type nseRegistryService struct {
+	clusterInfoService clusterinfo.ClusterInfoServer
+}
+
+func newNseRegistryService(clusterInfoService clusterinfo.ClusterInfoServer) *nseRegistryService {
+	return &nseRegistryService{
+		clusterInfoService: clusterInfoService,
+	}
+}
+
+func (rs *nseRegistryService) RegisterNSE(ctx context.Context, request *registry.NSERegistration) (*registry.NSERegistration, error) {
+	span := spanhelper.FromContext(ctx, "ProxyNsmgr.RegisterNSE")
+	defer span.Finish()
+	logger := span.Logger()
+
+	logger.Infof("%s: received RegisterNSE(%v)", NSRegistryForwarderLogPrefix, request)
+
+	nsmrsURL := os.Getenv(NSMRSAddressEnv)
+	if strings.TrimSpace(nsmrsURL) == "" {
+		err := errors.Errorf("NSMRS Address variable was not set")
+		logger.Warnf("%s: Skipping Register NSE forwarding: %v", NSRegistryForwarderLogPrefix, err)
+		return request, err
+	}
+
+	nodeConfiguration, cErr := rs.clusterInfoService.GetNodeIPConfiguration(ctx, &clusterinfo.NodeIPConfiguration{NodeName: request.NetworkServiceManager.Name})
```

Comment:
Please use span.Context() to pass it to child method to properly build a tree of spans.